File: android/src/main/java/com/azifacemobile/models/ProcessorResponse.java
package com.azifacemobile.models;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;

import org.json.JSONObject;

public class ProcessorResponse {
  private final WritableMap object;

  public ProcessorResponse() {
    this.object = Arguments.createMap();

    this.reset();
  }

  private void parseJSON(JSONObject target) {
    final ProcessorData processorData = new ProcessorData(target);

    final WritableMap data = processorData.getMap();
    final String responseBlob = data.getString("responseBlob");

    this.object.putBoolean("isSuccess", responseBlob != null);
    this.object.putMap("data", data);
    this.object.putMap("error", null);
  }

  public Boolean getSuccess() {
    return this.object.getBoolean("isSuccess");
  }

  public ReadableMap getData() {
    return this.object.getMap("data");
  }

  public void setData(JSONObject target) {
    this.parseJSON(target);
  }

  public WritableMap getMap() {
    return this.object.copy();
  }

  public void setError(String message, String code) {
    this.reset();

    final ProcessorError processorError = new ProcessorError(message, code);

    this.object.putMap("error", processorError.getMap());
  }

  public void reset() {
    this.object.putBoolean("isSuccess", false);
    this.object.putMap("data", null);
    this.object.putMap("error", null);
  }
}

File: ios/Models/ProcessorResponse.swift
import Foundation

public class ProcessorResponse {
  private var object: [AnyHashable: Any?]
  
  init() {
    self.object = [:]
    
    self.reset()
  }
  
  func parseJSON(target: [AnyHashable: Any]) {
    let processorData = ProcessorData(target: target)
    
    let data = processorData.getDict()
    let responseBlob = data.value(forKey: "responseBlob")
    
    self.object["isSuccess"] = responseBlob != nil
    self.object["data"] = data
    self.object["error"] = nil
  }
  
  func getSuccess() -> Bool {
    let isSuccess = self.object["isSuccess"]
    if (isSuccess == nil) {
      return false
    }
    
    return isSuccess as! Bool
  }
  
  func getData() -> NSDictionary? {
    let data = self.object["data"]
    if (data == nil) {
      return nil
    }
        
    return NSDictionary(dictionary: data as! [AnyHashable: Any])
  }
  
  func setData(target: [AnyHashable: Any]) {
    self.parseJSON(target: target)
  }
  
  func getDict() -> NSDictionary {
    return NSDictionary(dictionary: self.object as [AnyHashable: Any])
  }
  
  func setError(message: String, code: String) {
    self.reset()
    
    let processorError = ProcessorError(message: message, code: code)
    
    self.object["error"] = processorError.getDict()
  }
  
  func reset() {
    self.object["isSuccess"] = false
    self.object["data"] = nil
    self.object["error"] = nil
  }
}
